### PR TITLE
feat(Observable): unhandled errors are now reported to HostReportErrors

### PR DESCRIPTION
**Pull request by benlesh**
BREAKING CHANGE: Unhandled errors are no longer caught and rethrown, rather they are caught and scheduled to be thrown, which causes them to be reported to window.onerror or process.on('error'), depending on the environment. Consequently, teardown after a synchronous, unhandled, error will no longer occur, as the teardown would not exist, and producer interference cannot occur

### CODE REVIEW

**Comment by kwonoj** [comment]
Will check this soon.

File: src/Subscriber.ts [hunk 1]
```diff
@@ -199,80 +191,49 @@ class SafeSubscriber<T> extends Subscriber<T> {
...
     if (!this.isStopped && this._next) {
-      const { _parentSubscriber } = this;
-      if (!_parentSubscriber.syncErrorThrowable) {
-        this.__tryOrUnsub(this._next, value);
-      } else if (this.__tryOrSetError(_parentSubscriber, this._next, value)) {
+      try {
+        this._next.call(this._context, value);
+      } catch (err) {
         this.unsubscribe();
+        this._hostReportError(err);
```
Comment (kwonoj) [inline src/Subscriber.ts]: not sure if this need to be aligned - cases like error / complete we do report error first then unsubscribe, for next we do unsubscribe first then report. Nitpick maybe?

File: src/operators/tap.ts [hunk 2]
```diff
@@ -74,49 +76,62 @@ class DoOperator<T> implements Operator<T, T> {
...
-
-    const safeSubscriber = new Subscriber<T>(nextOrObserver, error, complete);
-    safeSubscriber.syncErrorThrowable = true;
-    this.add(safeSubscriber);
-    this.safeSubscriber = safeSubscriber;
-  }
+      super(destination);
+      this._tapError = error || noop;
+      this._tapComplete = complete || noop;
+      this._tapNext = noop;
```
Comment (kwonoj) [inline src/operators/tap.ts]: 83 already assigned into noop, do we need double init?

**Comment by kwonoj** [comment]
Except few nits I think change looks ok, probably biggest thing we'd like to get opinion is this breaking change itself I guess.

File: src/Subscriber.ts (hunk 1 above)
Comment (benlesh) [reply src/Subscriber.ts]: The error is reported asynchronously, so in this case it doesn't matter. However, I think it's probably supposed to be report error then unsubscribe. I'll fix it.

**Comment by benlesh** [comment]
> opinion is this breaking change itself I guess.

The opinions have already been expressed [here](https://github.com/tc39/proposal-observable/issues/119)

It's part of the spec, and frankly I really am not interested in this discussion again. We want to align with the proposal.

File: src/operators/tap.ts (hunk 2 above)
Comment (benlesh) [reply src/operators/tap.ts]: Fixed.

File: src/Subscriber.ts (hunk 1 above)
Comment (benlesh) [reply src/Subscriber.ts]: Improved.

**Comment by martinsik** [comment]
I feel it's hard for me to imagine what is this change going to cause (for example in Angular apps). I guess it shouldn't change basically anything for me because I don't rely on rethrowing errors but I'm not completely sure.

Maybe this change should be described in [MIGRATION.md](https://github.com/ReactiveX/rxjs/blob/master/MIGRATION.md) with some before/after example.

**Pull request merged**